File: Calculator/Calculator.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <stdlib.h>

#define MAX_HISTORY 100

void clearInputBuffer() {
    int c;
    while ((c = getchar()) != '\n' && c != EOF);
}

void tampilHistory(double history[], int count) {
    printf("History:\n");
    for (int i = 0; i < count; i++) {
        printf("%d. %.2lf\n", i + 1, history[i]);
    }
}

double hitung(double a, double b, char op[]) {
    if (strcmp(op, "+") == 0) return a + b;
    if (strcmp(op, "-") == 0) return a - b;
    if (strcmp(op, "*") == 0) return a * b;
    if (strcmp(op, "/") == 0) {
        if (b != 0) return a / b;
        printf("Error! Tidak bisa dibagi nol.\n");
        return 0;
    }
    if (strcmp(op, "**") == 0) return pow(a, b);

    printf("Operator tidak valid!\n");
    return 0;
}

int main() {
    double a, b, hasil;
    char op[10];
    double history[MAX_HISTORY];
    int count = 0;

    printf("\n=== Kalkulator ===\n");

    // looping input
    while (1) {
        printf("\nInput (Operasi, sqrt = √(akar), exit = selesai): ");

        // input keluar atau sqrt
        scanf("%s", op);

        if (strcmp(op, "exit") == 0) break;

        if (strcmp(op, "sqrt") == 0) {
            scanf("%lf", &a);
            if (a >= 0) {
                hasil = sqrt(a);
            } else {
                printf("Error! Akar negatif.\n");
                continue;
            }
        } else {
            // input dari awal string > double
            char *end; // konversi ke double
            a = strtod(op, &end);

                if (*end != '\0') {
                    printf("Input tidak valid!\n");
                    clearInputBuffer();
                    continue;
                }                                       
            scanf("%s %lf", op, &b);

            // validasi operator
            if (strcmp(op, "+") != 0 &&
                strcmp(op, "-") != 0 &&
                strcmp(op, "*") != 0 &&
                strcmp(op, "/") != 0 &&
                strcmp(op, "**") != 0) {
                printf("Operator tidak valid!\n");
                 clearInputBuffer();
                continue;
            }

            hasil = hitung(a, b, op);
        }

        printf("Hasil: %.2lf\n", hasil);

        if (count < MAX_HISTORY) {
            history[count++] = hasil;
        }

        tampilHistory(history, count);
    }

    printf("Program selesai.\n");
    return 0;
}

File: Calculatorpython/calculator.py
import math

def hitung(a, b, op):
    if op == '+':
        return a + b
    elif op == '-':
        return a - b
    elif op == '*':
        return a * b
    elif op == '/':
        if b != 0:
            return a / b
        else:
            return "Error: pembagian nol"
    elif op == '**':
        return a ** b
    else:
        return "Error: operator tidak valid"

with open("ops.txt", "r") as infile, open("results.txt", "w") as outfile:
    for line in infile:
        line = line.strip()

        if not line:
            continue

        parts = line.split()

        # sqrt
        if parts[0] == "sqrt":
            try:
                a = float(parts[1])
                if a >= 0:
                    hasil = math.sqrt(a)
                else:
                    hasil = "Error: akar negatif"
            except:
                hasil = "Error: input tidak valid"

        # operasi biasa
        else:
            try:
                a = float(parts[0])
                op = parts[1]
                b = float(parts[2])
                hasil = hitung(a, b, op)
            except:
                hasil = "Error: format salah"

        outfile.write(f"{line} = {hasil}\n")

print("Selesai! Cek results.txt")